internal/controller/controllers/agentserviceconfig_controller_test.go

```diff
@@ -0,0 +1,189 @@
+)
+
+func newTestReconciler(initObjs ...runtime.Object) *AgentServiceConfigReconciler {
+	c := fakeclient.NewFakeClientWithScheme(scheme.Scheme, initObjs...)
+	return &AgentServiceConfigReconciler{
+		Client:    c,
+		Scheme:    scheme.Scheme,
+		Log:       ctrl.Log.WithName("testLog"),
+		Namespace: testNamespace,
+	}
+}
+
+var _ = Describe("agentserviceconfig_controller reconcile", func() {
+	var (
+		asc      *aiv1beta1.AgentServiceConfig
+		ascr     *AgentServiceConfigReconciler
+		ctx      = context.Background()
+		mockCtrl *gomock.Controller
+		route    *routev1.Route
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		asc = newDefaultAgentServiceConfig()
+		ascr = newTestReconciler(asc)
+
+		// AgentServiceConfig created route is missing Host.
+		// We create one here with a value set for Host so that
+		// reconcile does not fail.
+		route, _ = ascr.newAgentRoute(asc)
+		route.Spec.Host = "testHost"
+		Expect(ascr.Client.Create(ctx, route)).To(BeNil())
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		Expect(ascr.Client.Delete(ctx, asc)).To(Succeed())
+		Expect(ascr.Client.Delete(ctx, route)).To(Succeed())
+	})
+
+	It("reconcile should succeed", func() {
+		result, err := ascr.Reconcile(newAgentServiceConfigRequest(asc))
+		Expect(err).To(BeNil())
+		Expect(result).To(Equal(ctrl.Result{}))
+	})
+})
+
+var _ = Describe("ensureAgentLocalAuthSecret", func() {
+	var (
+		asc             *aiv1beta1.AgentServiceConfig
+		ascr            *AgentServiceConfigReconciler
+		ctx             = context.Background()
+		mockCtrl        *gomock.Controller
+		privateKey      = "test-private-key"
+		publicKey       = "test-public-key"
+		localAuthSecret *corev1.Secret
+	)
+
+	BeforeEach(func() {
+		mockCtrl = gomock.NewController(GinkgoT())
+		asc = newDefaultAgentServiceConfig()
+		ascr = newTestReconciler(asc)
+	})
+
+	AfterEach(func() {
+		mockCtrl.Finish()
+		Expect(ascr.Client.Delete(ctx, asc)).To(Succeed())
+	})
+
+	Context("with an existing local auth secret", func() {
+		It("should not modify existing keys", func() {
+			localAuthSecret = &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      agentLocalAuthSecretName,
+					Namespace: testNamespace,
+				},
+				StringData: map[string]string{
+					"ec-private-key.pem": privateKey,
+					"ec-public-key.pem":  publicKey,
+				},
+				Type: corev1.SecretTypeOpaque,
+			}
+			Expect(ascr.Client.Create(ctx, localAuthSecret)).To(Succeed())
+
+			err := ascr.ensureAgentLocalAuthSecret(ctx, asc)
+			Expect(err).To(BeNil())
+
+			found := &corev1.Secret{}
+			err = ascr.Client.Get(context.TODO(), types.NamespacedName{Name: agentLocalAuthSecretName, Namespace: testNamespace}, found)
+			Expect(err).To(BeNil())
+
+			Expect(found.StringData["ec-private-key.pem"]).To(Equal(privateKey))
+			Expect(found.StringData["ec-public-key.pem"]).To(Equal(publicKey))
+
+			Expect(ascr.Client.Delete(ctx, localAuthSecret)).To(Succeed())
```

Comment:
Should this be moved into the `AfterEach` block? It looks like the second test doesn't remove the secret.

Reply:
Alternatively if the fake client is using an in-memory store and we're creating a new client object for each example then I think this line can just be removed.

Asked about this in my last review comment.

Reply:
I reviewed the fake client implementation, and like you said, the client is just tracking the objects in memory and since a new client is created for each test, object cleanup is not necessary. I'll remove the deletes.